gems/sorbet-runtime/lib/types/private/sealed.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+# typed: false
+
+module T::Private::Sealed
+  module NoInherit
+    def inherited(other)
+      super
+
+      this_line = Kernel.caller.find {|line| !line.match(/in `inherited'/)}
+      this_file = this_line&.split(':').first
+      decl_file = self.instance_variable_get(:@sorbet_sealed_module_decl_file)
+
+      if !this_file || !decl_file
+        raise "Couldn't determine enough file information for checking sealed classes"
+      end
+
+      if !this_file.start_with?(decl_file)
+        raise "#{self} was declared sealed and can only be inherited in #{decl_file}, not #{this_file}"
+      end
+    end
+  end
+
+  module NoIncludeExtend
+    def included(other)
+      super
+      this_line = Kernel.caller.find {|line| !line.match(/in `included'/)}
+      this_file = this_line&.split(':').first
+      decl_file = self.instance_variable_get(:@sorbet_sealed_module_decl_file)
+
+      if !this_file || !decl_file
+        raise "Couldn't determine enough file information for checking sealed modules"
+      end
+
+      if !this_file.start_with?(decl_file)
+        raise "#{self} was declared sealed and can only be included in #{decl_file}, not #{this_file}"
+      end
+    end
+
+    def extended(other)
+      super
+      this_line = Kernel.caller.find {|line| !line.match(/in `extended'/)}
+      this_file = this_line&.split(':').first
+      decl_file = self.instance_variable_get(:@sorbet_sealed_module_decl_file)
```

Comment:
Is it worth putting this logic into a helper function?

Reply:
Thinking about it. It's hard, because using a helper function changes the `caller` stack.